### ISSUE TITLE
problem with emscripten::val working with an array containing a Symbol

### ISSUE THREAD

**Issue by robfors** [emscripten-core/emscripten#7555]
I called a _JavaScript_ function with a `emscripten::val`, passing an array as one of the arguments. During the call, a `Symbol` was added to the array and number returned. I obviously expected the number to be returned to CPP, however, I got an error instead.

It produced the error `TypeError: Cannot convert a Symbol value to a string` from [_emval_equals](https://github.com/kripken/emscripten/blob/b6b2c0ff37ed761a3ea8ad0371bf985f5004cfde/src/embind/emval.js#L277).
This error was the result of comparing the array with a string I had previously send or returned. I didn't bother to look into why `emscripten` needed to do the comparison, but I'm sure that part is fine.

Next, I tired to get the error on my own. Sure enough, running
```js
[Symbol()] == 1
```
in a browser console resulted in the same error. To me this seems like a bug with _ECMAScript_ rather than anything else. I can run
```js
[] == 1
```
and
```js
Symbol() == 1
```
without errors, but when I put the `Symbol` in the array all of a sudden there is a problem? :smiley:

I get the result with _Chromium_ and _Firefox_ and it's hard to imagine that all those developers over looked this behavior.

Should I submit a PR where the comparison in `_emval_equals` is protected in a `try` block?

**Comment by chadaustin** [comment]
Wow, that JavaScript behavior is pretty surprising indeed. I'm not sure a `try` block in `_emval_equals` is a good idea, though. It might have a runtime performance impact. I know previous JS JITs would not optimize functions with try/catch in them.

**Comment by embeddedt** [comment]
I came across this issue and found the behavior you described rather interesting (I can still reproduce it on the latest Chrome). So I decided to do some experimenting.

**TL;DR**

[Symbols can't be implicitly cast to strings](https://stackoverflow.com/a/40944739).

**Long version**

Like you, I found that `[Symbol()] == 1` throws a `TypeError`. I also tried `[Symbol()] === 1` (three equals signs) which worked and just returned `false`. I presumed this was because `===` does not attempt to compare across type boundaries (it just bails out and returns `false`).

Since the common type between an array and a number is most likely a string, I then tried running `[Symbol()].toString()` which, again, gave the same `TypeError`.

Interestingly enough, running `Symbol().toString()` still worked fine. At this point I decided to do some research and came across the Stack Overflow post I linked above.

My assumption (from interpreting the backtrace of the `TypeError`) is that under the hood the implementation of `Array.toString` tries to implicitly convert each of the elements to a `string`. This would explain why `Symbol().toString()` works but `[Symbol()].toString()` doesn't.

**Comment by brion** [comment]
Yes, [per spec Array.prototype.toString](https://tc39.es/ecma262/#sec-array.prototype.tostring) calls the [`join` method](https://tc39.es/ecma262/#sec-array.prototype.join) which explicitly converts all the members to strings with the [internal ToString operation](https://tc39.es/ecma262/#sec-tostring) which is specified to throw a `TypeError` on Symbol input.

This is imo unfortunate but unlikely to be changed by the spec, so we should be careful with auto-coercing comparisons... Note that any JS object with a `toString` method could could cause similar comparison failures:

```js
var obj = {
    toString: function() {
        throw new TypeError('whoops');
    }
}
var arr = [obj];
arr == 1;
```

```
$ node foo.js 
/Users/brion/src/barfo2/foo.js:3
        throw new TypeError('whoops');
        ^

TypeError: whoops
    at Object.toString (/Users/brion/src/barfo2/foo.js:3:15)
    at Array.toString (native)
    at Object.<anonymous> (/Users/brion/src/barfo2/foo.js:7:5)
    at Module._compile (internal/modules/cjs/loader.js:776:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:787:10)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
    at Function.Module.runMain (internal/modules/cjs/loader.js:829:12)
    at startup (internal/bootstrap/node.js:283:19)
```
